### ISSUE TITLE
Anchor validation false positives after upgrading to 1.6

### ISSUE THREAD

**Issue by squidfunk** [mkdocs/mkdocs#3690]
First of all, thanks for the very useful [new anchor validation feature](https://www.mkdocs.org/about/release-notes/#anchors)! Upgrading to MkDocs 1.6, I was able to [fix many stale anchors in our documentation](https://github.com/squidfunk/mkdocs-material/issues/7076#issuecomment-2076300125). However, in the process, I came across a few false positives – the anchors are correct, but MkDocs reports them nonetheless. This is particularly the case when using the [Tabbed](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/) extension with [`combine_header_slug`](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/#tab-ids), which allows to prepend the slug of the nearest header to the tab.

I believe this needs to be fixed in MkDocs, since it might be some kind of ordering issue when running anchor validation, but I'm tagging @facelessuser here for visibility. By using the [Tabbed](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/) extension, anchor links to tabs are much more readable, something that many users asked for on the Material for MkDocs issue tracker.

## Example

The following example is used for illustration. You can also download it here: [anchor-validation-bug.zip](https://github.com/mkdocs/mkdocs/files/15144456/anchor-validation-bug.zip)

`index.md`

``` markdown
# Page with tabs

## Headline

=== "Tab A"

    Anchor for this tab is `#headline-tab-a`

=== "Tab B"

    Anchor for this tab is `#headline-tab-b`
```

`foo.md`

``` markdown
# Links to tabs

- [Tab A](index.md#headline-tab-a)
- [Tab B](index.md#headline-tab-b)
```

MkDocs will not detect those links and report them as non-existent:

```
INFO    -  Cleaning site directory
INFO    -  Doc file 'foo.md' contains a link 'index.md#headline-tab-a', but the doc 'index.md' does not contain an anchor '#headline-tab-a'.
INFO    -  Doc file 'foo.md' contains a link 'index.md#headline-tab-b', but the doc 'index.md' does not contain an anchor '#headline-tab-b'.
INFO    -  Documentation built in 0.12 seconds
```

When running the example, it's easy to see that anchor are correct and work as expected.

**Comment by pawamoy** [comment]
pymdownx Tabbed seems to register its treeprocessor updating anchors (slugs) with priority 4.

https://github.com/facelessuser/pymdown-extensions/blob/c8856a3bca3840b538ae3d03298a7d6ecf279541/pymdownx/tabbed.py#L416

...while MkDocs registers its treeprocessor reading anchors with priority 5.

https://github.com/mkdocs/mkdocs/blob/fb1d106747930b094bc464189d4b1c349fc541c6/mkdocs/structure/pages.py#L343

It means MkDocs gets anchors before Tabbed had a chance to update them. To fix this, either Tabbed should increase its treeprocessor priority, or MkDocs should lower its own.

**Comment by joapuiib** [comment]
Same has happend to me while using [Caption Block Pymdown Extension](https://facelessuser.github.io/pymdown-extensions/extensions/blocks/plugins/caption/)

- https://github.com/facelessuser/pymdown-extensions/issues/2529

**Comment by facelessuser** [comment]
> It means MkDocs gets anchors before Tabbed had a chance to update them. To fix this, either Tabbed should increase its treeprocessor priority, or MkDocs should lower its own (or both). The MkDocs priority has to be strictly lower than pymdownx's.

I just saw this. MkDocs, if they are going to validate anchors, needs to ensure they check as late as possible. There may be reasons a plugin requires to do things later. If MkDocs wants to be reliable, they should plan for worst case. I will have to check if there is an explicit reason I chose the priority I did. But if I can do earlier, its a small thing I can do.

But keep in mind, Markdown extensions don't care about MkDocs, or better put, they may not be designed with MkDocs in mind, MkDocs should be defensive of this and hold off as long as they can. MkDocs is not the only consumer of Python Markdown, and many users may not be aware of or use MkDocs.

**Comment by facelessuser** [comment]
It looks like Tabs is taking extra effort to produce unique IDs after  Toc has resolved its unique IDs, that's why we chose what we did. This is precisely why MkDocs should wait until as late as possible. I think this is indeed an MkDocs issue. Caption does not take the same pains (not currently at least) to produce unique IDs, but it may in the future.

**Comment by pawamoy** [comment]
Thanks @facelessuser! We should have tagged you earlier :smile: Sounds good, I note this should be fixed in MkDocs directly :slightly_smiling_face:

**Comment by facelessuser** [comment]
To clarify in case there is any debate on this point, when we introduced IDs in Tabs being created from titles, there was a risk of duplicate IDs, and additionally, duplicate IDs with existing headers.

If we chose a priority before Toc, when someone adds Tabbed, to an existing project, we increase the chance that existing header IDs may change due to a Tab creating an ID before Toc does, then Toc must de-duplicate their ID. The choice was made that Tab IDs would be lower priority. This means Tab would allow Toc to keep whatever titles it resolves, and we would de-duplicate ours. We felt this is less hostile to users.

**Comment by feasgal** [comment]
@squidfunk, would this same fix address the anchor validation false positives I'm getting on links to a [code selection](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#code-selection-button)? I was about to raise an issue over on your github, but if it's the same fix, I'll just follow this issue.

**Comment by squidfunk** [comment]
@feasgal only if you select single lines, and not multiple lines. Multi-line selection requires JavaScript (GitHub does the same on their code blocks) to resolve to the starting and ending anchors. Example:

- https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#__codelineno-4-2 (single, actual anchor)
- https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#__codelineno-4-2:3 (multi, not a real anchor)

I think it's a good idea to discuss whether MkDocs should allow to ignore some anchors during validation, i.e., those that are generated and consumed in JavaScript. For Material or MkDocs, it's only code selection, but there might be other themes impacted.

**Comment by squidfunk** [comment]
@pawamoy You're right, this should probably be handled as a separate issue. However, I thought it might be useful to start a discussion first, whether this would actually be considered an issue with MkDocs, or if we need to work around it.

**Comment by pdmosses** [comment]
@squidfunk [wrote](https://github.com/mkdocs/mkdocs/issues/3690#issuecomment-2617797331):

> Multi-line selection requires JavaScript

However, [text fragment directives](https://developer.mozilla.org/en-US/docs/Web/URI/Reference/Fragment/Text_fragments) allow selection of multiple lines – even across block boundaries – _without_ JavaScript.

For example, the following URL selects the text between the (first) occurrence of the specified start text and the first _subsequent_ occurrence of the word "text" in the target page:
https://developer.mozilla.org/en-US/docs/Web/URI/Reference/Fragment/Text_fragments#:~:text=Otherwise%20known%20as%20the%20fragment%20directive%2C%20this%20sequence%20of%20characters%20tells%20the%20browser%20that%20what%20comes%20next%20is%20one%20or%20more%20user%2Dagent%20instructions,text

> I think it's a good idea to discuss whether MkDocs should allow to ignore some anchors during validation, i.e., those that are generated and consumed in JavaScript.

It would be great if MkDocs could validate text fragment directives! Many characters need to be [percent-encoded](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent), which can make the anchors very difficult to read.

I think validation of text fragment directives (for local pages) should check whether the specified fragments exist, and emit an INFO showing the percent-_decoded_ anchor for each non-matching directive.

However, I will probably have about 100 text fragment links in a (Material for) MkDocs site that I'm developing, and `mkdocs serve` (v1.6.1) reports all such links each time I update the local preview. The many false positives will make it very difficult to see any genuine anchor validation issues. I don't want to set _all_ anchor validation to be ignored in `mkdocs.yml`; using the `--quiet` option suppresses also warnings (and even prevents the `--strict` option from aborting the build on _warnings_, it seems!).

Ideally, MkDocs would implement accurate anchor validation for text fragment directives. Until then, my preference would be to add a setting to ignore validation of those anchors. The current INFO reports are in any case misleading, e.g.:

```
INFO    -  Doc file 'SOURCE.md' contains a link 'TARGET.md#:~:text=7.2.', but
           the doc 'TARGET.md' does not contain an anchor '#:~:text=7.2.'.
```

> You're right, this should probably be handled as a separate issue.

Indeed – but it might still be good to take text fragment anchors into account when addressing the original issue.

**Comment by squidfunk** [comment]
@pdmosses you're missing the point – I was specifically talking about selection of lines in code blocks as a response to https://github.com/mkdocs/mkdocs/issues/3690#issuecomment-2617122005. Please note that this question is not related to the issues that is discussed here.

**Comment by pawamoy** [comment]
I have added the "Contributions welcome" and "Easy" labels: the suggestion solution is to decrease the treeprocessor processor priority to 3.9:

https://github.com/mkdocs/mkdocs/blob/fb1d106747930b094bc464189d4b1c349fc541c6/mkdocs/structure/pages.py#L343
